build(aio): auto-fill width/height to all image tags

petebacondarwin opened the pull request:
Parse all `<img>` tags, during doc-gen, and insert the width and height of
the sourceed image, if neither are already specified.

Errors are reported if the `<img>` tag has no `src` attribute or the image
cannot be loaded.

The work is done in the `addImageDimensions` processor, which must be
configured with a `basePath` so that it knows where to find the images.

petebacondarwin committed:
build(aio): move attribute utils to helpers folder

This allows these utility functions to be reused across packages.

petebacondarwin commented:
So we have a few missing images:

```
info:    running processor: addImageDimensions
error:   Unable to load src in image tag `<img src="assets/images/guide/dependency-injection/logged-in-user.png" alt="Logged In User">` - doc "guide/dependency-injection-in-action" (content)  - from file "guide/dependency-injection-in-action.md" - starting at line 1
error:   Unable to load src in image tag `<img src="assets/images/guide/dependency-injection/hero-bios.png" alt="Bios">` - doc "guide/dependency-injection-in-action" (content)  - from file "guide/dependency-injection-in-action.md" - starting at line 1
error:   Unable to load src in image tag `<img src="assets/images/guide/dependency-injection/hero-bio-and-content.png" alt="bio and contact">` - doc "guide/dependency-injection-in-action" (content)  - from file "guide/dependency-injection-in-action.md" - starting at line 1
error:   Unable to load src in image tag `<img src="assets/images/guide/dependency-injection/hero-bios-and-contacts.png" alt="Bios with contact into">` - doc "guide/dependency-injection-in-action" (content)  - from file "guide/dependency-injection-in-action.md" - starting at line 1
error:   Unable to load src in image tag `<img src="assets/images/guide/dependency-injection/hero-bio-contact-no-host.png" alt="Without @Host">` - doc "guide/dependency-injection-in-action" (content)  - from file "guide/dependency-injection-in-action.md" - starting at line 1
error:   Unable to load src in image tag `<img src="assets/images/guide/dependency-injection/highlight.png" alt="Highlighted bios">` - doc "guide/dependency-injection-in-action" (content)  - from file "guide/dependency-injection-in-action.md" - starting at line 1
error:   Unable to load src in image tag `<img src="assets/images/guide/dependency-injection/minimal-logger-intellisense.png" alt="MinimalLogger restricted API">` - doc "guide/dependency-injection-in-action" (content)  - from file "guide/dependency-injection-in-action.md" - starting at line 1
error:   Unable to load src in image tag `<img src="assets/images/guide/dependency-injection/sorted-heroes.png" alt="Sorted Heroes">` - doc "guide/dependency-injection-in-action" (content)  - from file "guide/dependency-injection-in-action.md" - starting at line 1
error:   Unable to load src in image tag `<img src="assets/images/guide/dependency-injection/alex.png" alt="Alex in action">` - doc "guide/dependency-injection-in-action" (content)  - from file "guide/dependency-injection-in-action.md" - starting at line 1
error:   Unable to load src in image tag `<img src="assets/images/guide/dependency-injection/alice.png" alt="Alice in action">` - doc "guide/dependency-injection-in-action" (content)  - from file "guide/dependency-injection-in-action.md" - starting at line 1
error:   Unable to load src in image tag `<img src="assets/images/bios/stephenfluin.jpg">` - doc "news" (content)  - from file "news.html" - starting at line 1
error:   Unable to load src in image tag `<img src="assets/images/bios/igor-minar.jpg">` - doc "news" (content)  - from file "news.html" - starting at line 1
error:   Unable to load src in image tag `<img src="assets/images/bios/shield-bio-placeholder.png">` - doc "news" (content)  - from file "news.html" - starting at line 1
error:   Unable to load src in image tag `<img src="assets/images/bios/angular-gde-bio-placeholder.png">` - doc "news" (content)  - from file "news.html" - starting at line 1
error:   Unable to load src in image tag `<img src="assets/images/bios/shield-bio-placeholder.png">` - doc "news" (content)  - from file "news.html" - starting at line 1
error:   Unable to load src in image tag `<img src="assets/images/bios/shield-bio-placeholder.png">` - doc "news" (content)  - from file "news.html" - starting at line 1
error:   Unable to load src in image tag `<img src="/assets/images/logos/angularjs/AngularJS-Shield.svg" alt="AngularJS Logo" style="margin-left:20px;" >` - doc "presskit" (content)  - from file "presskit.html" - starting at line 1
```

petebacondarwin commented:
Should we remove the `news` page altogether? @wardbell

petebacondarwin committed:
build(aio): include new packages in docs-watch

gkalpak commented:
FYI, I have fixed many of the missing images in #16318.

mary-poppins commented:
The angular.io preview for 8d68a8728d50f4b5e49ba28c93bca85f98b7c6e2 is available [here][1].

[1]: https://pr16353-8d68a8728d50f4b5e49ba28c93bca85f98b7c6e2.ngbuilds.io/

mary-poppins commented:
The angular.io preview for d67fb610a31d18e4fe8675839acb84ed0f8cf19f is available [here][1].

[1]: https://pr16353-d67fb610a31d18e4fe8675839acb84ed0f8cf19f.ngbuilds.io/

petebacondarwin commented:
I am going to rebuild this using my new "post-processing" processor :-)

wardbell reviewed (approved):
I see you're doing it for both MD and HTML images now. Let's merge it (although @gkalpak is right, you should fix that comment).

petebacondarwin committed:
build(aio): move copyContentAssets processor to the base package

This allows other processors who need to know about the copyContentAssets
processors to ensure that the runs after the content has been copied.

petebacondarwin committed:
build(aio): capture and log errors and warnings when post-processing HTML

petebacondarwin committed:
style(aio): fix comment on postProcessHtml processor

Fixes https://github.com/angular/angular/pull/16336#discussion_r113891332

mary-poppins commented:
The angular.io preview for cacafc52bdde5dc98e76814493bb3f7d1fbd7e7b is available [here][1].

[1]: https://pr16353-cacafc52bdde5dc98e76814493bb3f7d1fbd7e7b.ngbuilds.io/

petebacondarwin committed:
build(aio): auto-fill width/height to all image tags

Parse all `<img>` tags, during doc-gen, and insert the width and height of
the sourceed image, if neither are already specified.

Warnings are reported if the `<img>` tag has no `src` attribute or the image
cannot be loaded.

The work is done in the `addImageDimensions` post-processor, which must be
configured with a `basePath` so that it knows where to find the images.

Closes #15888

petebacondarwin committed:
docs(aio): fix image size on homepage

mary-poppins commented:
The angular.io preview for 9acf4c3b665c53719487272145c26504db389fde is available [here][1].

[1]: https://pr16353-9acf4c3b665c53719487272145c26504db389fde.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 75a3f81a0d389d61acd37f99ffcfe0f0301f4c1e is available [here][1].

[1]: https://pr16353-75a3f81a0d389d61acd37f99ffcfe0f0301f4c1e.ngbuilds.io/

petebacondarwin commented on aio/tools/transforms/angular-base-package/processors/copyContentAssets.js:
```diff
@@ -1,7 +1,6 @@
 
 module.exports = function copyContentAssetsProcessor(copyFolder) {
   return {
-    $runAfter: ['files-written'],
```
at this stage it doesn't matter when this processor runs, since it is independent of any other processing.

petebacondarwin commented on aio/tools/transforms/angular-base-package/services/copyFolder.js:
```diff
@@ -0,0 +1,5 @@
+const {copySync} = require('fs-extra');
+
+module.exports = function copyFolder() {
+  return (from, to) => copySync(from, to);
```
switched from `shelljs/cp` to `fs-extra/copySync` because the former breaks if the destination folder does not yet exist.

mary-poppins commented:
The angular.io preview for 1ee4b873b282c4060a923d699d5b8473e16c252e is available [here][1].

[1]: https://pr16353-1ee4b873b282c4060a923d699d5b8473e16c252e.ngbuilds.io/

wardbell commented on aio/tools/transforms/authors-package/api-package.js:
```diff
@@ -10,17 +10,18 @@ const apiPackage = require('../angular-api-package');
 const { API_SOURCE_PATH } = require('../config');
 
 const packageMap = {
+  animations: ['animations/index.ts', 'animations/browser/index.ts', 'animations/browser/testing/index.ts'],
```
While seemingly unrelated to the PR purpose these are well worth adding and I have no objection to slipping them in here.

wardbell commented on aio/tools/transforms/post-process-package/processors/post-process-html.js:
```diff
@@ -9,15 +9,15 @@ const rehype = require('rehype');
  * Each plugin is a factory function that will be called with the "rehype" engine as `this`.
- * The factory should return a function that takes a HAST and returns a `boolean` or `undefined`.
- * The HAST can be mutated by the this function.
+ * The factory should return a "transform"" function that takes a HAST and returns a `boolean` or `undefined`.
```
The editor gave you double double-quotes: `""`.

wardbell reviewed (changes requested):
Just a little "oops" with the comment on a`add-image-dimensions.js` and then it seems LGTM.

wardbell commented on aio/tools/transforms/angular-base-package/post-processors/add-image-dimensions.js:
```diff
@@ -0,0 +1,38 @@
+const source = require('unist-util-source');
+
+/**
+ * Add in the document path to links that start with a hash.
```
This seems like a comment copied from a galaxy far far away.

mary-poppins commented:
The angular.io preview for 17772800fc5bcea0d71c96d47fd590cc6e8d9369 is available [here][1].

[1]: https://pr16353-17772800fc5bcea0d71c96d47fd590cc6e8d9369.ngbuilds.io/

wardbell commented:
The current travis failures appear to be perf test failures that are unrelated to this PR which should be mergeable.
